### ISSUE TITLE
inlineRequires: true not working on default exports/imports

### ISSUE THREAD

**Issue by robrechtme** [react/metro#909]
**Do you want to request a *feature* or report a *bug*?**
Bug

**What is the current behavior?**

When setting `inlineRequires` to `true`, I'd expect all imports to be transformed to inline requires. However, it seems like this is not working for default exports. Check out the following example: 

```js
// metro.config.js
module.exports = {
  transformer: {
    getTransformOptions: async () => ({
      transform: {
        experimentalImportSupport: false,
        inlineRequires: true,
      },
    }),
  },
};
```
```js
// defaultFn.js
console.log('defaultFn.js imported');

export default function defaultFn() {
  console.log('defaultFn executed');
}
```

```js
// namedFn.js
console.log('namedFn.js imported');

export function namedFn() {
  console.log('namedFn executed');
}
```
```jsx
// index.js
import React from 'react';
import { AppRegistry, Button, SafeAreaView } from 'react-native';
import { name } from './app.json';
import { namedFn } from './namedFn';
import defaultFn from './defaultFn';

function App() {
  const onPress = () => {
    namedFn();
    defaultFn();
  };

  return (
    <SafeAreaView>
      <Button title="Click me" onPress={onPress} />
    </SafeAreaView>
  );
}

AppRegistry.registerComponent(name, () => App);
```

When launching the app, I immediately get the following logs:
```
defaultFn.js imported
```
After pressing the button, the following is logged:
```
namedFn.js imported
namedFn executed
defaultFn executed
```

**What is the expected behavior?**

I would expect the `defaultFn.js imported` log to also occur only after pressing the button. The log indicates that the `defaultFn.js` file is imported and parsed, and `inlineRequires: true` is not behaving correctly? Or is this by design?

**Please provide your exact Metro configuration and mention your Metro, node, yarn/npm version and operating system.**
- react-native 0.70.6
- node 16.16.0
- npm 8.16.0
- MacOS

**Comment by motiz88** [comment]
Thanks @robrechtme! This is a known issue. `experimentalImportSupport: true` solves this, but the option is currently undocumented and experimental (as reflected in its name). We're looking into what it would involve to make it the default behaviour in Metro.

**Comment by robrechtme** [comment]
Hi @motiz88

Are you sure? Changing `experimentalImportSupport` has no effect on my example. 

I'm using `react-native@0.70.6` which resolves metro to:
```
    metro-react-native-babel-transformer "0.72.3"
    metro-runtime "0.72.3"
    metro-source-map "0.72.3"
```

**Comment by bitcrumb** [comment]
Hi @motiz88,

I'm seeing the same as @robrechtme.

Both `experimentalImportSupport` and `inlineRequires` are set to true, but the default exports are never inlined. I also verify this by setting a `console.log()` top-level in a component file.

When default exporting the component, the `console.log()` is always executed immediately (so no inline require happens). But when "regularly" exporting the component, the `console.log()` is only executed when rendered.

This is rather annoying, since most of our components are exported as defaults.

I even converted all our default exports to regular exports, just to be able to profit from this feature 😅

**Comment by motiz88** [comment]
@robrechtme Is there a `babel.config.js` file in your project? One quirk of the current implementation of `experimentalImportSupport: true` is that it depends on disabling Babel's ES Module support, which you may need to do manually at the moment.

**Comment by bitcrumb** [comment]
There is in ours, but it's rather limited:

```
module.exports = function babelConfig(api) {
  api.cache.using(() => process.env.NODE_ENV);

  const presets = ['module:metro-react-native-babel-preset'];

  const plugins = ['react-native-reanimated/plugin'];

  if (!(api.env('development') || api.env('test'))) {
    plugins.push('transform-remove-console');
  }

  plugins.push([
    require.resolve('babel-plugin-module-resolver'),
    {
      root: ['./src/'],
      alias: {
...
      },
    },
  ]);

  return {
    presets,
    plugins,
  };
};
```

How exactly do I disable Babel ES Module support?

**Comment by bitcrumb** [comment]
👀

**Comment by robrechtme** [comment]
I could reproduce this in a fresh RN app with `npx react-native init`, there the babel config is:

```js
// babel.config.js
module.exports = {
  presets: ['module:metro-react-native-babel-preset'],
};
```

**Comment by aemre** [comment]
How fix it via custom resolver?

**Comment by Rag0n** [comment]
> I even converted all our default exports to regular exports, just to be able to profit from this feature (hundreds of components) 😅

Hey @bitcrumb , did you measure improvements? We also often use default export, I'm wondering how beneficial it is to switch to a named export.

**Comment by robhogan** [comment]
`experimentalImportSupport` looked like a no-op because you have a `babel.config.js`. `@react-native/metro-babel-transformer` only applies `@react-native/babel-preset` with its own options - including `disableImportExportTransform: experimentalImportSupport` - when the project has no Babel config of its own. If one exists it simply extends it, so Babel's CommonJS transform keeps running regardless of the Metro flag. See https://github.com/facebook/react-native/blob/main/packages/react-native-babel-transformer/src/index.js#L92-L118.

That's what breaks the inlining. With the CommonJS transform in play, `import defaultFn from './defaultFn'` compiles to a module-scope `_interopRequireDefault(require('./defaultFn'))`, and the inline-requires plugin only inlines bare `require` calls. Named imports compile to a plain `require(...).namedFn`, which is why only those were being inlined.

So set both - `experimentalImportSupport: true` in your Metro config, and the matching preset option in `babel.config.js`:

```js
module.exports = {
  presets: [
    ['module:@react-native/babel-preset', {disableImportExportTransform: true}],
  ],
};
```

Transformed output, before and after:

```js
// inlineRequires: true
var _defaultFn = _interopRequireDefault(require(_dependencyMap[1], "./defaultFn"));
function App() {
  (0, require(_dependencyMap[2], "./namedFn").namedFn)();
  (0, _defaultFn.default)();
}

// inlineRequires + experimentalImportSupport + disableImportExportTransform
function App() {
  require(_dependencyMap[0], "./namedFn").namedFn();
  _$$_IMPORT_DEFAULT(_dependencyMap[1], "./defaultFn")();
}
```

Closing on that basis. If it's still not inlining on a current Metro with both set, please open a new issue with the transformed output.
